### ISSUE TITLE
Can't share volume with host 

### ISSUE THREAD

**Issue by zmsp** [GoogleCloudPlatform/kubernetes#12278]
Hi, I have been trying to share volume with host. I have tried creating a nfs server as well as hostPath but it doesn't seem to allow me to see the files on host inside pod container. Here is my pod file 

```
apiVersion: v1
kind: Pod
metadata:
  name: postgresql2
spec:
  containers:
  - image: postgres:latest
    name: postgres
  ports:
  - containerPort: 5432
    hostPort: 5432
    name: pg-port
  volumeMounts:
    - mountPath: /test-pd
      name: test-volume
      readOnly: false
volumes:
  - hostPath:
    name: test-volume
    path: /export
``` 
Thanks.

**Comment by erictune** [comment]
I had to play around for a bit to figure out what was wrong with your spec, so we definitely need to improve our error reporting.  

The problem is that the "volumes" block needs to be indented to be under "spec".
Also, ports and volumeMounts need to be indented under "containers".

The following should work:
```
apiVersion: v1
kind: Pod
metadata:
  name: postgresql2
spec:
  containers:
  - image: postgres:latest
    name: postgres
    ports:
    - containerPort: 5432
      hostPort: 5432
      name: pg-port
    volumeMounts:
      - mountPath: /test-pd
        name: test-volume
        readOnly: false
  volumes:
    - hostPath:
      name: test-volume
      path: /export
```

**Comment by erictune** [comment]
Actually, even that is not right.  Also the "path" needs to be nested under "hostPath".  Edited above comment.

**Comment by erictune** [comment]
Ways I debugged this:

`kubectl get pods/postgresql2 -o yaml` and then check if all the fields in the original yaml are also in the returned yaml.    Each time the yaml was wrong, there was a field missing from the returned yaml.  Fields with the wrong indentation are silently ignored.  

Create pod with ``kubectl create --validate -f somefile.yaml`.
When there were mistakes in the pod, it printed error like these:
```
I0805 10:43:25.129850   46757 schema.go:126] unknown field: path
I0805 10:43:25.129973   46757 schema.go:129] this may be a false alarm, see https://github.com/GoogleCloudPlatform/kubernetes/issues/6842
pods/host-path-example
```

**Comment by erictune** [comment]
feel free to reopen if still not working.

**Comment by erictune** [comment]
Added docs to help other users in  #12296.

**Comment by zmsp** [comment]
Thanks! Well explained

**Comment by erictune** [comment]
did it work?

**Comment by zmsp** [comment]
Yes

**Comment by swagiaal** [comment]
> kubectl get pods/postgresql2 -o yaml and then check if all the fields in the original yaml are also in the returned yaml. Each time the yaml was wrong, there was a field missing from the returned yaml. Fields with the wrong indentation are silently ignored.

Neat!

**Comment by zmsp** [comment]
Hi, I was able to resolve my issue of sharing a NFS volume using the commands you provided. However hostPath volume still doesn't work. Debugging the code doesn't seem to throw any error but I don't see shared volume files that are on container on my host volume. Here is my code.

```
apiVersion: v1
kind: Pod
metadata:
  name: postgresql
spec:
  containers:
  - image: postgres:latest
    name: postgres
    ports:
    - containerPort: 5432
      hostPort: 5432
      name: pg-port
    volumeMounts:
     - mountPath: /var/lib/postgresql
       name: nfs
  volumes:
    - name: nfs
      hostPath:
        path: /export
```

**Comment by yunshengb** [comment]
I posted my yaml file in http://stackoverflow.com/questions/37819488/kubernetes-hostpath-volume-does-not-mount

Can anyone solve my issue there? I really do not know why my certificate files are not mounted into pod. Thanks!

**Comment by somoza** [comment]
Probably... You're mounting the NFS on the master node... and it need to be mounted in the nodes/minions...

**Comment by fracasula** [comment]
Hi there, I'm having some issues sharing a volume on a Debian system with minikube. It works well on a bunch of Macs I have here in the office (tried on at least 5 different machines) but I just can't make it work on that Debian machine.

The deployment file is [this one here](https://github.com/fracasula/k8s-test/blob/develop/deploy/nodeclient_deployment.yaml).
I checked with VirtualBox and the mounted folder in Debian is `/home` rather than `/Users` so I changed that accordingly both in the `nodeclient_deployment.yaml` file and in the [Makefile](https://github.com/fracasula/k8s-test/blob/develop/Makefile) (which I use to run `make mount`, a shortcut for creating/symlinking the folder I want to be mounted).

So it doesn't seem to be working and I don't see any particular output when mounting or creating deployments. Any idea on how I can debug this?

**Note:** both VirtualBox setups on Linux and OSX are using the latest version (ext pack included).
